internal/metrics/reporter.go

```diff
@@ -0,0 +1,326 @@
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base32"
+	"fmt"
+	"io"
+	mathrand "math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/lightstep/lightstep-tracer-common/golang/gogo/collectorpb"
+	"github.com/lightstep/lightstep-tracer-common/golang/gogo/metricspb"
+)
+
+const (
+	DefaultReporterAddress             = "https://ingest.lightstep.com:443"
+	DefaultReporterTimeout             = time.Second * 5
+	DefaultReporterMeasurementDuration = time.Second * 30
+)
+
+var (
+	acceptHeader      = http.CanonicalHeaderKey("Accept")
+	contentTypeHeader = http.CanonicalHeaderKey("Content-Type")
+	accessTokenHeader = http.CanonicalHeaderKey("Lightstep-Access-Token")
+)
+
+const (
+	reporterPath = "/metrics"
+
+	idempotencyKeyByteLength = 30
+	protoContentType         = "application/octet-stream"
+
+	ReporterPlatformKey        = "lightstep.reporter_platform"
+	ReporterPlatformVersionKey = "lightstep.reporter_platform_version"
+	ReporterVersionKey         = "lightstep.reporter_version"
+)
+
+type Reporter struct {
+	client               *http.Client
+	tracerID             uint64
+	attributes           map[string]string
+	address              string
+	timeout              time.Duration
+	accessToken          string
+	stored               Metrics
+	collectorReporter    *collectorpb.Reporter
+	labels               []*collectorpb.KeyValue
+	Start                time.Time
+	End                  time.Time
+	MetricsCount         int
+	skippedInitialReport bool
+}
+
+func attributesToTags(attributes map[string]string) []*collectorpb.KeyValue {
+	tags := []*collectorpb.KeyValue{}
+	for k, v := range attributes {
+		tags = append(tags, &collectorpb.KeyValue{Key: k, Value: &collectorpb.KeyValue_StringValue{StringValue: v}})
+	}
+	return tags
+}
+
+func getLabels(attributes map[string]string) []*collectorpb.KeyValue {
+	labels := []*collectorpb.KeyValue{}
+	filters := []string{
+		"lightstep.component_name",
```

Comment:
nit: use the constants from `options.go` (or is there a circular dependency? ...argh)

Reply:
it was a circular dep, i added a `constants` package to move out at least the common keys